From 116fe838c19b29727f7677b3cf00ecad0c832da7 Mon Sep 17 00:00:00 2001
From: Julian Brown <julian.brown@cpanel.net>
Date: Tue, 3 Sep 2024 15:22:03 -0500
Subject: [PATCH] libxml2 2.13 makes changes to how the parsing state is set,
 update our code accordingly. In particular, it started reporting entities
 within attributes, while it should only report entities inside text nodes.

---
 ext/xml/compat.c           |  2 +-
 ext/xml/tests/gh14834.phpt | 29 +++++++++++++++++++++++++++++
 2 files changed, 30 insertions(+), 1 deletion(-)
 create mode 100644 ext/xml/tests/gh14834.phpt

diff --git a/ext/xml/compat.c b/ext/xml/compat.c
index 18a7beae..3de77d07 100644
--- a/ext/xml/compat.c
+++ b/ext/xml/compat.c
@@ -376,7 +376,7 @@ _get_entity(void *user, const xmlChar *name)
 			ret = xmlGetDocEntity(parser->parser->myDoc, name);
 
 		ZEND_DIAGNOSTIC_IGNORED_START("-Wdeprecated-declarations")
-		if (ret == NULL || (parser->parser->instate != XML_PARSER_ENTITY_VALUE && parser->parser->instate != XML_PARSER_ATTRIBUTE_VALUE)) {
+		if (ret == NULL || parser->parser->instate == XML_PARSER_CONTENT) {
 		ZEND_DIAGNOSTIC_IGNORED_END
 			if (ret == NULL || ret->etype == XML_INTERNAL_GENERAL_ENTITY || ret->etype == XML_INTERNAL_PARAMETER_ENTITY || ret->etype == XML_INTERNAL_PREDEFINED_ENTITY) {
 				/* Predefined entities will expand unless no cdata handler is present */
diff --git a/ext/xml/tests/gh14834.phpt b/ext/xml/tests/gh14834.phpt
new file mode 100644
index 00000000..2781ba2e
--- /dev/null
+++ b/ext/xml/tests/gh14834.phpt
@@ -0,0 +1,29 @@
+--TEST--
+GH-14834 (Error installing PHP when --with-pear is used)
+--EXTENSIONS--
+xml
+--FILE--
+<?php
+$xml = <<<XML
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE root [
+    <!ENTITY foo "ent">
+]>
+<root>
+  <element hint="hello&apos;world">&foo;<![CDATA[ &amp; ]]><?x &amp; ?></element>
+</root>
+XML;
+
+$parser = xml_parser_create();
+xml_set_character_data_handler($parser, function($_, $data) {
+    var_dump($data);
+});
+xml_parse($parser, $xml, true);
+?>
+--EXPECT--
+string(3) "
+  "
+string(3) "ent"
+string(7) " &amp; "
+string(1) "
+"
-- 
2.48.2

